### PR TITLE
mgr/dashboard: fix bootstrap script for cephadm installation

### PR DESCRIPTION
**Pull request by avanthakkar**
Recent refactoring of cephadm files broke the cephadm installation for dashboard cephadm e2e jenkins jobs.
More description about issue: https://github.com/ceph/ceph/pull/53052#issuecomment-1713162764.
The PR fixes this issue by generating cephadm binaries under `ceph_dev_folder` src and symlink to `/root/bin/cephadm`

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/pybind/mgr/dashboard/ci/cephadm/bootstrap-cluster.sh [hunk 1]
```diff
@@ -8,8 +8,13 @@ mkdir /root/bin
 export CEPHADM_IMAGE='quay.ceph.io/ceph-ci/ceph:main'
 
 CEPHADM="/root/bin/cephadm"
+CEPHADM_SRC="/mnt/{{ ceph_dev_folder }}/src/cephadm/cephadm"
+
+ls -l /mnt/{{ ceph_dev_folder }}/src/cephadm/
+chmod 777 /mnt/{{ ceph_dev_folder }}/src/cephadm/
+/mnt/{{ ceph_dev_folder }}/src/cephadm/build.sh $CEPHADM_SRC
+ln -s $CEPHADM_SRC $CEPHADM
```
[the remark this answers is not in the document]
Comment (avanthakkar) [reply src/pybind/mgr/dashboard/ci/cephadm/bootstrap-cluster.sh]: Makes sense, thanks! Pushed the changes, PTAL

[the remark this answers is not in the document]
Comment (nizamial09) [reply src/pybind/mgr/dashboard/ci/cephadm/bootstrap-cluster.sh]: don't see the changes though?

File: src/pybind/mgr/dashboard/ci/cephadm/bootstrap-cluster.sh [hunk 2]
```diff
@@ -8,8 +8,11 @@ mkdir /root/bin
 export CEPHADM_IMAGE='quay.ceph.io/ceph-ci/ceph:main'
 
 CEPHADM="/root/bin/cephadm"
+CEPHADM_SRC="/mnt/{{ ceph_dev_folder }}/src/cephadm/cephadm"
+
+# /mnt/{{ ceph_dev_folder }}/src/cephadm/build.sh $CEPHADM_SRC
+ln -s $CEPHADM_SRC $CEPHADM
 
-/mnt/{{ ceph_dev_folder }}/src/cephadm/build.sh $CEPHADM
 mkdir -p /etc/ceph
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/ci/cephadm/bootstrap-cluster.sh:16]: i don't think changes in this file is necessary. we can probably keep it as it is and it should work

File: src/pybind/mgr/dashboard/ci/cephadm/start-cluster.sh [hunk 3]
```diff
@@ -63,6 +63,8 @@ cd ${CEPH_DEV_FOLDER}
 : ${VM_IMAGE_URL:='https://download.fedoraproject.org/pub/fedora/linux/releases/36/Cloud/x86_64/images/Fedora-Cloud-Base-36-1.5.x86_64.qcow2'}
 kcli download image -p ceph-dashboard -u ${VM_IMAGE_URL} ${VM_IMAGE}
 kcli delete plan -y ceph || true
+# Compile cephadm locally for the shared_ceph_folder to pick it up
+${CEPH_DEV_FOLDER}/src/cephadm/build.by ${CEPH_DEV_FOLDER}/src/cephadm/cephadm
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/ci/cephadm/start-cluster.sh:67]: we are trying to execute a python file here
```suggestion
exec python3 ${CEPH_DEV_FOLDER}/src/cephadm/build.by ${CEPH_DEV_FOLDER}/src/cephadm/cephadm
```

File: src/pybind/mgr/dashboard/ci/cephadm/bootstrap-cluster.sh [hunk 4]
```diff
@@ -8,8 +8,10 @@ mkdir /root/bin
 export CEPHADM_IMAGE='quay.ceph.io/ceph-ci/ceph:main'
 
 CEPHADM="/root/bin/cephadm"
+CEPHADM_SRC="/mnt/{{ ceph_dev_folder }}/src/cephadm/cephadm"
+
+ln -s $CEPHADM_SRC $CEPHADM
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/ci/cephadm/bootstrap-cluster.sh:13]:
```suggestion
cp /mnt/{{ ceph_dev_folder }}/src/cephadm/cephadm $CEPHADM
```

**Review by nizamial09 — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard/ci/cephadm/bootstrap-cluster.sh (hunk 4 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/ci/cephadm/bootstrap-cluster.sh:13]: not sure if a copy would ask for sudo permission in jenkins!

**Comment by avanthakkar** [comment]
jenkins test dashboard cephadm

File: src/cephadm/cephadm.py [hunk 5]
```diff
@@ -2486,8 +2486,11 @@ def get_container_mounts(ctx, fsid, daemon_type, daemon_id,
         if ctx.shared_ceph_folder:  # make easy manager modules/ceph-volume development
             ceph_folder = pathify(ctx.shared_ceph_folder)
             if os.path.exists(ceph_folder):
+                cephadm_binary = ceph_folder + '/src/cephadm/cephadm'
+                if not os.path.exists(pathify(cephadm_binary)):
+                    raise Error("cephadm binary does not exist. Please run './build.sh cephadm' from ceph/src/cephadm/ directory.")
+                mounts[cephadm_binary] = '/usr/sbin/cephadm'
```
Comment (adk3798) [inline src/cephadm/cephadm.py:2492]: FYI, Redouane's PR got merged so you should be able to rebase and remove this commit

**Comment by avanthakkar** [comment]
jenkins test dashboard cephadm

**Review by nizamial09 — APPROVED** [review]

**Comment by nizamial09** [comment]
jenkins test api

**Comment by nizamial09** [comment]
Thanks @avanthakkar

**Pull request merged**